### PR TITLE
net: lib: Library to parse iCalendar

### PR DESCRIPTION
**Pull request by lats1980**
iCalendar(https://tools.ietf.org/html/rfc5545) is data format
for exchanging calendar information. The iCalendar parser library
can be used to parse a data stream with iCalendar format.

When the stream is parsed, the library report the parsed component
with errors code. The library supports below calendar components
and properties:

Calendar components:
    * VEVENT

Calendar properties:
    * DTSTART
    * DTEND
    * SUMMARY
    * LOCATION
    * DESCIPTION

Signed-off-by: Larry Tsai <larry.tsai@nordicsemi.no>

### CODE REVIEW

**Comment by lats1980** [comment]
The original pull request was messed up due to force pushing local repo.
https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1946

The changes reflects to the reviewer's comment was in this pull request.

File: samples/nrf9160/icalendar_parser/src/main.c [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_SEC_TAG_LIST, sec_tag_list,
+			 sizeof(sec_tag_t) * ARRAY_SIZE(sec_tag_list));
+	if (err) {
+		printk("Failed to set socket security tag, errno %d\n", errno);
+		return -1;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(verify));
```
Comment (rlubos) [inline samples/nrf9160/icalendar_parser/src/main.c]: Duplicated peer verify option?

File: samples/nrf9160/icalendar_parser/src/main.c [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+	}
+}
+
+static int socket_sectag_set(int fd, int sec_tag)
+{
+	int err;
+	int verify;
+	sec_tag_t sec_tag_list[] = { sec_tag };
+
+	enum {
```
Comment (rlubos) [inline samples/nrf9160/icalendar_parser/src/main.c]: I've added appropriate defines into `socket.h` some time ago, let's use them instead:
https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/master/include/net/socket.h#L126

File: include/net/icalendar_parser.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * @brief iCalendar parser event.
+ */
+struct ical_parser_evt {
+	/** Event ID. */
+	enum ical_parser_evt_id id;
+	union {
+		/** Error cause. */
+		int error;
```
[the remark this answers is not in the document]
Comment (lats1980) [reply include/net/icalendar_parser.h]: Originally the error event reports those socket connection errors. Now the error event is used to report parsing error such as
 - unsupported components
 - property value overflow
 - wrong calendaring format

Not added yet. I will add such error report to the next commit.

File: samples/nrf9160/icalendar_parser/src/main.c (hunk 1 above)
Comment (lats1980) [reply samples/nrf9160/icalendar_parser/src/main.c]: Removed.

File: include/net/icalendar_parser.rst [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+################
+
+`iCalendar`_ is data format for representing and exchanging calendaring and scheduling information. The iCalendar parser library parses data steam containg iCalendar format calendaring information  and returns paresd calendar events.
+
+The library first detects the begin of calendar delimiter(BEGIN:VCALENDAR) and parses following calendar content fragment by fragment and sends parsed events (:cpp:member:`ICAL_PARSER_EVT_COMPONENT`) to the application when calendar components are parsed.
+
+Protocols
+*********
+
+The library supports HTTP and HTTPS (TLS 1.2) over IPv4 and IPv6.
```
[the remark this answers is not in the document]
Comment (lats1980) [reply include/net/icalendar_parser.rst]: Rewrite description to:

The library supports iCalendar version 2.0 defined in RFC5545.

File: include/net/icalendar_parser.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	char dtstart[CONFIG_ICAL_PARSER_DTSTART_SIZE];
+	/** DTEND property buffer. */
+	char dtend[CONFIG_ICAL_PARSER_DTEND_SIZE];
+};
+
+/**
+ * @brief iCalendar parser asynchronous event handler.
+ *
+ * Through this callback, the application receives events, such as
+ * parsed component, download completion, or errors.
```
[the remark this answers is not in the document]
Comment (lats1980) [reply include/net/icalendar_parser.h]: Removed. No download behavior now.

File: include/net/icalendar_parser.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (lats1980) [reply include/net/icalendar_parser.h]: Now the error code is reported along with parsing event.

/**
 * @brief iCalendar parser event IDs.
 */
enum ical_parser_evt_id {
	/** Event Component parsed */
	ICAL_EVT_VEVENT,
	/** Todo Component parsed */
	ICAL_EVT_VTODO,
	/** Journal Component parsed */
	ICAL_EVT_VJOURNAL,
	/** Timezone Component parsed */
	ICAL_EVT_VTIMEZONE,
	/** FreeBusy Component parsed */
	ICAL_EVT_VFREEBUSY,
};

/**
 * @brief iCalendar parser error IDs.
 */
enum ical_parser_error_id {
	/** No error */
	ICAL_ERROR_NONE,
	/** SUMMARY property error */
	ICAL_ERROR_SUMMARY,
	/** LOCATION property error */
	ICAL_ERROR_LOCATION,
	/** DESCRIPTION property error */
	ICAL_ERROR_DESCRIPTION,
	/** DTSTART property error */
	ICAL_ERROR_DTSTART,
	/** DTEND property error */
	ICAL_ERROR_DTEND,
	/** Component not supported error */
	ICAL_ERROR_COM_NOT_SUPPORTED,
};

File: samples/nrf9160/icalendar_parser/src/main.c (hunk 2 above)
Comment (lats1980) [reply samples/nrf9160/icalendar_parser/src/main.c]: Modified per the definition.

**Comment by lats1980** [comment]
Calendar parsing sample is removed in this PR. The sample will be pushed to separated PR.

File: subsys/net/lib/icalendar_parser/src/icalendar_parser.c [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+
+static bool parse_desc_props(const char *buf,
+			     const char *name,
+			     size_t name_size,
+			     char *value,
+			     size_t max_value_len)
+{
+	bool ret;
+	size_t unfold_size;
+	char ical_prop_buf[CONFIG_ICAL_PARSER_MAX_PROPERTY_SIZE + 1];
```
Comment (lemrey) [inline subsys/net/lib/icalendar_parser/src/icalendar_parser.c]: That's 1K that is pushed on the stack right here, that's a lot. Do we need 1K on average, and have you considered using a static buffer instead?

File: subsys/net/lib/icalendar_parser/src/icalendar_parser.c [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+#include <strings.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <toolchain/common.h>
+#include <net/icalendar_parser.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(icalendar_parser, CONFIG_ICAL_PARSER_LOG_LEVEL);
+
+static size_t unfold_contentline(const char *buf, char *prop_buf)
```
Comment (lemrey) [inline subsys/net/lib/icalendar_parser/src/icalendar_parser.c]: It seems to me you're just trimming EOL sequences from the buffer here.
Does this function really need a separate output buffer, could it not modify the input buffer instead?

**Comment by lats1980** [comment]
@ru-fu , please review the document at path:
include/net/icalendar_parser.rst 

Many thanks.

File: include/net/icalendar_parser.rst [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+The library supports iCalendar version 2.0 as defined in `RFC5545 <iCalendar_>`_.
+
+The iCalendar parser library supports parsing the following calendar components:
+
+* VEVENT
+
+It supports parsing the following calendar properties:
+
+* DESCRIPTION
```
Comment (ru-fu) [inline include/net/icalendar_parser.rst:24]: I changed the order for these properties to be alphabetic. Is that okay, or should they be ordered differently?

File: include/net/icalendar_parser.rst [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+iCalendar parser
+################
+
+The iCalendar parser library can be used to parse a data stream in `iCalendar`_ format, which is a data format for representing and exchanging calendaring and scheduling information.
+The library parses the calendaring information and returns parsed calendar events.
+
+The library first detects the beginning of the calendar object by locating the delimiter ``BEGIN:VCALENDAR``.
+It then parses the following calendar content fragment by fragment.
+For each calendar component that is parsed, the library sends a parsed event (:c:type:`ical_parser_evt`) to the application.
```
Comment (ru-fu) [inline include/net/icalendar_parser.rst:11]: What is the object for the parsed event? I couldn't find ICAL_PARSER_EVT_COMPONENT in the code.

File: include/net/icalendar_parser.rst [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+It supports parsing the following calendar properties:
+
+* DESCRIPTION
+* DTEND
+* DTSTART
+* LOCATION
+* SUMMARY
+
+API documentation
```
Comment (ru-fu) [inline include/net/icalendar_parser.rst:30]: Are there any configuration options for the library that should be pointed out in the documentation? It seems to me like most users will be okay with the defaults and don't need to change the options. Correct?

File: include/net/icalendar_parser.rst (hunk 8 above)
Comment (lats1980) [reply include/net/icalendar_parser.rst:24]: It is totally ok. The order in RFC is also in alphabetic.

File: include/net/icalendar_parser.rst (hunk 10 above)
Comment (lats1980) [reply include/net/icalendar_parser.rst:30]: So far there are no configuration options. Adding config options for those optional properties like LOCATION/SUMMARY could improve memory usage, but I suggest to add options in the future.

File: include/net/icalendar_parser.rst (hunk 9 above)
Comment (lats1980) [reply include/net/icalendar_parser.rst:11]: ICAL_PARSER_EVT_COMPONENT is out of date. The event reported by library is now ical_parser_evt, which contains event id (ical_parser_evt_id), error code (ical_parser_error_id), and parsed component(struct ical_component ).

File: subsys/net/lib/icalendar_parser/Kconfig [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+menuconfig  ICAL_PARSER
+	bool "iCalendar parser"
+
+if ICAL_PARSER
+
+config ICAL_PARSER_BUFFER_SIZE
+	int "buffer size for unparsed data"
```
[the remark this answers is not in the document]
Comment (lats1980) [reply subsys/net/lib/icalendar_parser/Kconfig:13]: I will change the description of Kconfig.

File: subsys/net/lib/icalendar_parser/Kconfig [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+	default 2048
+	help
+	  Size for internal buffer to store unparsed data.
+
+config ICAL_PARSER_MAX_PROPERTY_SIZE
+	int "maximum size of iCalendar property"
```
[the remark this answers is not in the document]
Comment (lats1980) [reply subsys/net/lib/icalendar_parser/Kconfig:19]: I will change the description of Kconfig.

File: subsys/net/lib/icalendar_parser/Kconfig [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+	default 1024
+
+config ICAL_PARSER_SUMMARY_SIZE
+	int "maximum size of summart property"
```
[the remark this answers is not in the document]
Comment (lats1980) [reply subsys/net/lib/icalendar_parser/Kconfig:23]: I will change the description of Kconfig.

File: subsys/net/lib/icalendar_parser/Kconfig [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	range 32 ICAL_PARSER_MAX_PROPERTY_SIZE
+	default 64
+
+config ICAL_PARSER_LOCATION_SIZE
+	int "maximum size of location property"
```
[the remark this answers is not in the document]
Comment (lats1980) [reply subsys/net/lib/icalendar_parser/Kconfig:28]: I will change the description of Kconfig.

File: subsys/net/lib/icalendar_parser/Kconfig [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+	range 32 ICAL_PARSER_MAX_PROPERTY_SIZE
+	default 64
+
+config ICAL_PARSER_DESCRIPTION_SIZE
+	int "maximum size of description property"
```
[the remark this answers is not in the document]
Comment (lats1980) [reply subsys/net/lib/icalendar_parser/Kconfig:33]: I will change the description of Kconfig.

File: subsys/net/lib/icalendar_parser/Kconfig [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+	range 32 ICAL_PARSER_MAX_PROPERTY_SIZE
+	default 128
+
+config ICAL_PARSER_DTSTART_SIZE
+	int "maximum size of dtstart property"
```
[the remark this answers is not in the document]
Comment (lats1980) [reply subsys/net/lib/icalendar_parser/Kconfig:38]: I will change the description of Kconfig.

**Comment by lats1980** [comment]
> I added a commit - please check and squash if the updates look okay.
> Plus, I suggested a few language fixes to the Kconfig file.
> 
> Another question: Is "net" the correct folder for this library? In the documentation, that folder has the title "Libraries for networking", and this library doesn't seem to do any networking tasks.

Yes, the library does not do networking task now. It does parser task only. Is "Other libraries" a good place to put the library to?

**Comment by rlubos** [comment]
>>I added a commit - please check and squash if the updates look okay.
>>Plus, I suggested a few language fixes to the Kconfig file.

>>Another question: Is "net" the correct folder for this library? In the documentation, that folder >>has the title "Libraries for networking", and this library doesn't seem to do any networking tasks.

>Yes, the library does not do networking task now. It does parser task only. Is "Other libraries" a >good place to put the library to?

No, let's keep it in `net`. Merely the fact that the library itself does not exchange any packets does not mean it's not a networking library. Even the library name clearly implies it's networking related ("Internet calendar").

File: include/net/icalendar_parser.rst (hunk 9 above)
Comment (ru-fu) [reply include/net/icalendar_parser.rst:11]: OK, that's what I suggested as well, so we can probably leave this as is.

File: include/net/icalendar_parser.rst (hunk 10 above)
Comment (ru-fu) [reply include/net/icalendar_parser.rst:30]: ok :)

**Pull request merged**